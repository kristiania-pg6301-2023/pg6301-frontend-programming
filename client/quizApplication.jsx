import {HashRouter, Link, Route, Routes, useNavigate} from "react-router-dom";
import React, {useEffect, useState} from "react";

function FrontPage() {
  return (
    <>
      <h2>Welcome</h2>
      <Link to={"/question"}>Ask a question</Link>
    </>
  );
}

function QuestionAnswerButton({ answer, onClick }) {
  return (
    <div>
      <button onClick={onClick}>{answer}</button>
    </div>
  );
}

export function Question({ question, onClickAnswer }) {
  if (!question) {
      return <div>Loading...</div>;
  }
  return (
    <>
      <h2>Can you answer this?</h2>
      <p>{question.question}</p>
      {Object.keys(question.answers)
        .filter((k) => question.answers[k])
        .map((k) => (
          <QuestionAnswerButton
            key={k}
            answer={question.answers[k]}
            onClick={() => onClickAnswer(question.id, k)}
          />
        ))}
    </>
  );
}

function ShowAnswer({ onAskAnother }) {
  return (
    <>
      <Routes>
        <Route path={"/correct"} element={<h2>That's correct!</h2>} />
        <Route path={"/wrong"} element={<h2>That's wrong!</h2>} />
      </Routes>
      <div>
        <button onClick={onAskAnother}>Ask me another question</button>
      </div>
    </>
  );
}

function Quiz() {
  const [question, setQuestion] = useState();

  async function fetchRandomQuestion() {
      const res = await fetch("/api/questions/random");
      setQuestion(await res.json());
  }

  useEffect(() => {
      fetchRandomQuestion();
  }, [])

  const navigateFn = useNavigate();

  async function handleClickAnswer(id, answer) {
    const res = await fetch("/api/questions/answer", {
        method: "POST",
        body: JSON.stringify({id, answer}),
        headers: {
            "content-type": "application/json"
        }
    });
    const response = await res.json();
    if (response.correct) {
      navigateFn("/answer/correct");
    } else {
      navigateFn("/answer/wrong");
    }
  }

  async function handleAskAnother() {
    await fetchRandomQuestion();
    navigateFn("/question");
  }

  return (
    <Routes>
      <Route path={"/"} element={<FrontPage />} />
      <Route
        path={"/answer/*"}
        element={<ShowAnswer onAskAnother={handleAskAnother} />}
      />
      <Route
        path={"/question"}
        element={
          <Question question={question} onClickAnswer={handleClickAnswer} />
        }
      />
      <Route path={"*"} element={<h2>Not Found</h2>} />
    </Routes>
  );
}

export function QuizApplication() {
  return (
    <HashRouter>
      <header>
        <h1>React quiz app with express backend</h1>
      </header>
      <nav>
        <Link to={"/"}>Front page</Link>
      </nav>
      <main>
        <Quiz />
      </main>
      <footer>
        Created by Johannes Brodwall @ Kristiania University College
      </footer>
    </HashRouter>
  );
}
