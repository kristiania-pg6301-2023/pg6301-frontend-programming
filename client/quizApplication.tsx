import { HashRouter, Link, Route, Routes, useNavigate } from "react-router-dom";
import React, { useEffect, useState } from "react";

interface Question {
  id: number;
  question: string;
  answers: Record<string, string>;
}

interface Score {
  answers: number;
  correctAnswers: number;
}

function FrontPage() {
  return (
    <>
      <h2>Welcome</h2>
      <Link to={"/question"}>Ask a question</Link>
    </>
  );
}

function QuestionAnswerButton({
  answer,
  onClick,
}: {
  answer: string;
  onClick(): void;
}) {
  return (
    <div>
      <button onClick={onClick}>{answer}</button>
    </div>
  );
}

export function Question({
  question,
  onClickAnswer,
}: {
  question?: Question;
  onClickAnswer(id: number, k: string): void;
}) {
  if (!question) {
    return <div>Loading...</div>;
  }
  return (
    <>
      <h2>Can you answer this?</h2>
      <p>{question.question}</p>
      {Object.keys(question.answers)
        .filter((k) => question.answers[k])
        .map((k) => (
          <QuestionAnswerButton
            key={k}
            answer={question.answers[k]}
            onClick={() => onClickAnswer(question.id, k)}
          />
        ))}
    </>
  );
}

function ShowAnswer({ onAskAnother }: { onAskAnother(): void }) {
  return (
    <>
      <Routes>
        <Route path={"/correct"} element={<h2>That&apos;s correct!</h2>} />
        <Route path={"/wrong"} element={<h2>That&apos;s wrong!</h2>} />
      </Routes>
      <div>
        <button onClick={onAskAnother}>Ask me another question</button>
      </div>
    </>
  );
}

function ShowScore() {
  const [score, setScore] = useState<Score>();
  async function fetchScore() {
    const res = await fetch("/api/score");
    setScore(await res.json());
  }
  useEffect(() => {
    fetchScore();
  }, []);

  if (!score) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h2>
        Your have {score.correctAnswers} out of {score.answers} correct answers
      </h2>
    </>
  );
}

export function Quiz({
  fetchQuestion,
  postAnswer,
}: {
  fetchQuestion(): Promise<Question>;
  postAnswer(id: number, answer: string): Promise<{ correct: boolean }>;
}) {
  const [question, setQuestion] = useState<Question>();

  async function loadRandomQuestion() {
    setQuestion(await fetchQuestion());
  }

  useEffect(() => {
    loadRandomQuestion();
  }, []);

  const navigateFn = useNavigate();

  async function handleClickAnswer(id: number, answer: string) {
    const response = await postAnswer(id, answer);
    if (response.correct) {
      navigateFn("/answer/correct");
    } else {
      navigateFn("/answer/wrong");
    }
  }

  async function handleAskAnother() {
    await loadRandomQuestion();
    navigateFn("/question");
  }

  return (
    <Routes>
      <Route path={"/"} element={<FrontPage />} />
      <Route
        path={"/answer/*"}
        element={<ShowAnswer onAskAnother={handleAskAnother} />}
      />
      <Route
        path={"/question"}
        element={
          <Question question={question} onClickAnswer={handleClickAnswer} />
        }
      />
      <Route path={"/score"} element={<ShowScore />} />
      <Route path={"*"} element={<h2>Not Found</h2>} />
    </Routes>
  );
}

export function QuizApplication() {
  async function fetchQuestion() {
    const res = await fetch("/api/questions/random");
    return await res.json();
  }

  async function postAnswer(id: number, answer: string) {
    const res = await fetch("/api/questions/answer", {
      method: "POST",
      body: JSON.stringify({ id, answer }),
      headers: {
        "content-type": "application/json",
      },
    });
    return (await res.json()) as { correct: boolean };
  }

  return (
    <HashRouter>
      <header>
        <h1>React quiz app with express backend</h1>
      </header>
      <nav>
        <Link to={"/"}>Front page</Link>
        <Link to={"/score"}>See my score</Link>
      </nav>
      <main>
        <Quiz fetchQuestion={fetchQuestion} postAnswer={postAnswer} />
      </main>
      <footer>
        Created by Johannes Brodwall @ Kristiania University College
      </footer>
    </HashRouter>
  );
}
