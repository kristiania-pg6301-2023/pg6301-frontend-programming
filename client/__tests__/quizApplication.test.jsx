import renderer, {act} from "react-test-renderer";

import { Question } from "../quizApplication";

const sampleQuestion = {
  id: 1,
  question: "What's up",
  answers: {
    answer_a: "Nothing",
    answer_b: "Nothing much",
    answer_c: "'sup!"
  }
}

describe("quiz application", () => {

  it("shows a question", async () => {
    let component;
    await act(async () => {
      component = renderer.create(
        <Question question={sampleQuestion} onClickAnswer={jest.fn} />
      );
    });
    expect(component).toMatchSnapshot();
    expect(component.root.findByType("p").children.join(" ")).toEqual(sampleQuestion.question);
  })

})