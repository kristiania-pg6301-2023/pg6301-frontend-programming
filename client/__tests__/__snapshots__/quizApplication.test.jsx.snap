// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quiz application shows a question 1`] = `
[
  <h2>
    Can you answer this?
  </h2>,
  <p>
    What's up
  </p>,
  <div>
    <button
      onClick={[Function]}
    >
      Nothing
    </button>
  </div>,
  <div>
    <button
      onClick={[Function]}
    >
      Nothing much
    </button>
  </div>,
  <div>
    <button
      onClick={[Function]}
    >
      'sup!
    </button>
  </div>,
]
`;
